gs_chunked_io/async_collections.py

```diff
@@ -0,0 +1,125 @@
+"""
+Set-like and list-like collections to manage concurrent execution using ThreadPoolExecutor.
+"""
+import multiprocessing
+from concurrent.futures import Future, ThreadPoolExecutor, as_completed
+from typing import Any, List, Set, Callable, Generator, Optional
+
+
+class _AsyncCollection:
+    def __init__(self, executor: ThreadPoolExecutor, size: int=0):
+        self.executor = executor
+        self.size = size
+        assert 0 < self.size
+
+    def __len__(self):
+        raise NotImplementedError()
+
+    def __bool__(self) -> bool:
+        return bool(len(self))
+
+    def put(self, func, *args, **kwargs):
+        raise NotImplementedError()
+
+    def get(self) -> Any:
+        raise NotImplementedError()
+
+    def consume(self) -> Generator[Any, None, None]:
+        raise NotImplementedError()
+
+    def consume_finished(self) -> Generator[Any, None, None]:
+        raise NotImplementedError()
+
+    def _running(self) -> Set[Future]:
+        futures = getattr(self, "_futures", list())
+        return {f for f in futures if not f.done()}
+
+    def abort(self):
+        futures = getattr(self, "_futures", list())
+        for f in futures:
+            f.cancel()
+        for _ in as_completed(futures):
+            pass
+
+    def wait(self):
+        raise NotImplementedError()
+
+    def __del__(self):
+        self.abort()
+
+class AsyncSet(_AsyncCollection):
+    """
+    Set-like collection for future results, when order does not matter.
+    `submit` blocks until the number of in-progress operations becomes less than or equal to `size`.
+    """
+    def __init__(self, executor: ThreadPoolExecutor, size: int=multiprocessing.cpu_count()):
+        super().__init__(executor, size)
+        self._futures: Set[Future] = set()
+
+    def __len__(self):
+        return len(self._futures)
+
+    def put(self, func: Callable, *args, **kwargs):
+        running = len(self._running())
+        if running >= self.size:
+            self.wait(running - self.size)
+        f = self.executor.submit(func, *args, **kwargs)
+        self._futures.add(f)
+
+    def get(self) -> Any:
+        for r in self.consume():
+            return r
+
+    def consume(self) -> Any:
+        for f in as_completed(self._futures):
+            self._futures.remove(f)
+            yield f.result()
+
+    def consume_finished(self) -> Generator[Any, None, None]:
+        for f in self._futures.copy():
+            if f.done():
+                self._futures.remove(f)
+                yield f.result()
+
+    def wait(self, count: int=-1):
+        for i, f in enumerate(as_completed(self._running())):
+            f.result()
+            if i == count:
+                break
+
+class AsyncList(_AsyncCollection):
+    """
+    List-like collection for future results, when order of results matters.
```

Comment:
It's not immediately clear to me that this is intended to be used with a GET or a PUT operation or why, except from your PR description.  Could you add a little detail into the docstring about the intended usage?